### ISSUE TITLE
Ocean basin mask?

### ISSUE THREAD

**Issue by jbusecke** [mathause/regionmask#90]
I am working with ocean output of climate models. A common task is to mask out large scale ocean basins, like the Atlantic and then e.g. compute a zonal overturning stream-function.

I just started working on a first implementation using the [create your own region functionality](https://regionmask.readthedocs.io/en/stable/notebooks/create_own_regions.html#), but I think this could be a great addition to add to the package directly. I think many ocean scientists are looking for this kind of functionality.

Natural Earth seems to have ocean basins available [in the marine areas file](https://www.naturalearthdata.com/downloads/50m-physical-vectors/50m-physical-labels/).

If someone could provide me with some pointers I would be happy to try to submit a PR.

**Comment by mathause** [comment]
Happy to include this. Internally I use a function named `_obtain_ne`. 

``` python
import regionmask 
regionmask.defined_regions._obtain_ne
```

You can see examples here:

https://github.com/mathause/regionmask/blob/902e2738232ba5ba42cbd8d7867798d21b3b96c0/regionmask/defined_regions/natural_earth.py#L141-L152

**Comment by aaronspring** [comment]
I took shapefiles from https://www.marineregions.org/downloads.php

**Comment by jbusecke** [comment]
I have just implemented the marine areas from natural earth in #91. 
I am now able to do this locally:
```
import regionmask
import matplotlib.pyplot as plt
import numpy as np

def plot_grid(obj):
    plt.figure()
    lon = np.arange(-179.5, 179.5)
    lat = np.arange(89.5, -89.5, -1)
    mask = obj.mask(lon, lat, wrap_lon=False)
    mask_ma = np.ma.masked_invalid(mask)
    obj.plot(add_label=False, add_ocean=False);
    LON_EDGE = np.arange(-180, 180)
    LAT_EDGE = np.arange(90, -90, -1)
    plt.pcolormesh(LON_EDGE, LAT_EDGE, mask_ma, cmap='nipy_spectral');
    plt.tight_layout()
    return mask_ma

basins = regionmask.defined_regions.natural_earth.ocean_basins_50
plot_grid(basins)
```
I get the following plot:
![image](https://user-images.githubusercontent.com/14314623/77120562-85cbc880-6a0f-11ea-97ba-ee93153a6f74.png)

This is already great! But what I would really need is to merge all the marginal seas into the large basins. @mathause, does regionmask provide this kind of functionality? Or would that have to be integrated outside?

@aaronspring, does your data source provide the large basins **including** the seas? Then I would check if I can integrate that aswell.
Thanks for all the help.

**Comment by mathause** [comment]
@jbusecke Great, thanks, I'll have a look. Currently this is not implemented, but I'd be happy to add something. Just to clarify, would you like to (1) combine the regions for everyone (i.e. add a change to `obtain_ne`) or (2) load the original basins and combine them afterwards, e.g. something along the lines of

``` python
import regionmask
basins = regionmask.defined_regions.natural_earth.ocean_basins_50
basins_comb = basins.combine(pacific=["p_S", "p_N"])
```
?

**Comment by jbusecke** [comment]
That would be awesome! I think the latter provides the most flexibility. E.g. some folks might want to mask out the Chesapeake Bay, but for my purposes I want it to be part of the North Atlantic. I have hacked something together so I can work with the data now. 

```python
import warnings
def merged_mask(basins,ds, lon_name='lon', lat_name='lat', merge_dict = None):
    # Notes: Can I cache the basin data? Or is that done already?
    mask = basins.mask(ds,lon_name=lon_name, lat_name=lat_name)

    def find_mask_index(name):
        target_value = [ri for ri in range(len(basins.regions)) if basins.regions[ri].name == name]
        if len(target_value) > 1:
            warnings.warn(f"Found more than one matching region for {name}")
            return target_value[0]
        elif len(target_value) == 1:
            return target_value[0]
        else:
            return None
            
    
    if merge_dict is None:
        merge_dict = {
            'INDIAN OCEAN':['Mozambique Channel', 'Bay of Bengal', 'Arabian Sea', 'Persian Gulf', 'Andaman Sea',
                            'Laccadive Sea', 'Gulf of Aden', 'Gulf of Oman', 'Gulf of Mannar', 'Gulf of Kutch', 'Great Australian Bight'],
            'Arctic Ocean':['Beaufort Sea', 'Chukchi Sea', 'Barents Sea', 'Kara Sea', 'Laptev Sea', 'White Sea',
                            'The North Western Passages', 'Amundsen Gulf', 'Viscount Melville Sound'],
            'SOUTHERN OCEAN':['Ross Sea', 'Weddell Sea', 'Bellingshausen Sea', 'Amundsen Sea', 'Scotia Sea', 'Drake Passage'],
            'North Atlantic Ocean':['Caribbean Sea', 'Gulf of Mexico', 'Labrador Sea', 'Hudson Bay', 'Baffin Bay',
                                    'Norwegian Sea', 'Greenland Sea', 'Bay of Biscay', 'Norwegian Sea', 'Greenland Sea',
                                    'Gulf of Guinea', 'Irish Sea', 'North Sea', 'BahÃ\xada de Campeche', 'Davis Strait',
                                    'Sargasso Sea', 'Hudson Strait', 'English Channel', 'Gulf of Honduras', 'Bristol Channel',
                                    'Inner Seas', 'Straits of Florida','Gulf of Saint Lawrence', 'Bay of Fundy', 'Melville Bay',
                                    'Gulf of Maine', 'Chesapeake Bay', 'Amazon River', 'James Bay','Ungava Bay'],
            'North Pacific Ocean':['Philippine Sea',  'Gulf of Alaska', 'Sea of Okhotsk','East China Sea',
                                   'Yellow Sea', 'Bering Sea', 'Golfo de California', 'Korea Strait', 'Cook Inlet', 'Bristol Bay',
                                  'Shelikhova Gulf', 'Bo Hai', 'Golfo de PanamÃ¡', 'Yangtze River', 'Columbia River','Sea of Japan','Inner Sea'],
            'South Pacific Ocean':['Coral Sea', 'Tasman Sea','Bay of Plenty', 'Bismarck Sea', 'Solomon Sea',
                                   'Great Barrier Reef'],
            'South Atlantic Ocean':['RÃ\xado de la Plata', 'Golfo San Jorge'],
            'Black Sea':None,
            'Mediterranean Sea':['Tyrrhenian Sea' ,'Adriatic Sea', 'Golfe du Lion', 'Ionian Sea', 'Strait of Gibraltar', 'Balearic Sea', 'Aegean Sea',],
            'Red Sea':None,
            'Caspian Sea':None,
            'Maritime Continent':['Celebes Sea', 'Sulu Sea','Banda Sea', 'Luzon Strait', 'Java Sea', 'Arafura Sea',
                                  'Timor Sea', 'Gulf of Thailand', 'Gulf of Carpentaria', 'Molucca Sea', 'Gulf of Tonkin',
                                 'Strait of Malacca', 'Strait of Singapore', 'Makassar Strait', 'Ceram Sea', 'Taiwan Strait',
                                 'South China Sea',],
            'Baltic Sea':['Gulf of Bothnia', 'Gulf of Finland'],
                     }
    
    merged_basins = []
    for ocean, small_basins in merge_dict.items():
#         ocean_idx = find_mask_index(ocean)
        try:
            ocean_idx = basins.map_keys(ocean)
        except(KeyError):
            #The ocean key is new and cant be found in the previous keys (e.g. for Atlantic full or maritime continent)
            ocean_idx = mask.max().data + 1
            
        if small_basins:
            for sb in small_basins:
                sb_idx = basins.map_keys(sb)
                #set the index of each small basin to the ocean value
                mask = mask.where(mask!=sb_idx, ocean_idx)
                merged_basins.append(sb)
    remaining_basins = [str(basins.regions[ri].name) for ri in range(len(basins.regions)) if (basins.regions[ri].name not in merged_basins) and (basins.regions[ri].name not in list(merge_dict.keys()))]
    print(remaining_basins)
    
    #scale the numbers back to integers
    map_values = np.unique(mask.data)
    map_values = map_values[~np.isnan(map_values)]
    new_values = np.arange(len(map_values))
    for old, new in zip(map_values, new_values):
        mask = mask.where(mask!=old, new)
        
    return mask
```
As input dataset I use a CMIP6 model from the Google cloud (`conda install -c conda-forge gcsfs zarr intake-esm cmip6_preprocessing` is required)

```
import intake
from cmip6_preprocessing.preprocessing import combined_preprocessing
col_url = "https://raw.githubusercontent.com/NCAR/intake-esm-datastore/master/catalogs/pangeo-cmip6.json"
col = intake.open_esm_datastore(col_url)
cat = col.search(source_id='GFDL-ESM4', experiment_id='historical', variable_id='so', grid_label='gr')
data_dict = cat.to_dataset_dict(zarr_kwargs={'consolidated': True, 'decode_times': False})#,preprocess=combined_preprocessing
ds = data_dict['CMIP.NOAA-GFDL.GFDL-ESM4.historical.Omon.gr']
ds = combined_preprocessing(ds).isel(time=0).so.load()
ds
```

I can then do:
```
basins = regionmask.defined_regions.natural_earth.ocean_basins_50
mask = merged_mask(basins, ds)
```

This gives me:
![image](https://user-images.githubusercontent.com/14314623/77174310-3b8e2a00-6a97-11ea-8015-1cac63b34185.png)

That is already pretty awesome! I think with your implementation I can just keep that giant dictionary and apply it? It might be nice to provide this as a default somehow, but I can also maintain that myself. 

If you get to implement this, could you also check if it would be possible to combine regions with the same name? Right now the ne regions have two regions with the same name: The Mediterannean and the Ross Sea. In each case these are two distinct regions (You can see them in different colors in both seas in the map above). These should ideally be merged or renamed (I guess I could also go upstream to natural earth?)

**Comment by aaronspring** [comment]
including the seas?
I used lunghurst regions which I think didn’t include marginal seas like the Black Sea. Is it about whether this set of shapefiles covers the whole ocean?

https://www.marineregions.org/sources.php#longhurst

**Comment by aaronspring** [comment]
For longhurst I also aggregated many small inti larger areas.

**Comment by jbusecke** [comment]
> including the seas?

To clarify I was referring to the marginal seas, like e.g. the North Sea (which I grouped into the North Atlantic in the example above). Never heard of those lunghurst regions, but those look like a good alternative set tbh. At the moment I really only want a mask with North/South Atlantic, North/South Pacific, Indian, Southern Ocean and Arctic for my purposes.

**Comment by mathause** [comment]
Thanks for the example that looks promising! This could also be interesting to combine regions on a continent... I still need to think whether I would aggregate the regions before or after the masking and what to do with the name and the abbreviations...

**Comment by jbusecke** [comment]
I think if you can combine them before, and return a new object, that would be ideal. Let me know if you need any help with that. I assume this will be a separate PR. Do you think we need more work on #91?

**Comment by aaronspring** [comment]
> > including the seas?
> 
> To clarify I was referring to the marginal seas, like e.g. the North Sea (which I grouped into the North Atlantic in the example above). Never heard of those lunghurst regions, but those look like a good alternative set tbh. At the moment I really only want a mask with North/South Atlantic, North/South Pacific, Indian, Southern Ocean and Arctic for my purposes.

Maybe I can contribute my example based on shapefiles from that website with regionmask? But it got a bit ugly when the regions were too small for data in that gridbox and these things. I will provide my example next week and then we can see how we generalize this.

**Comment by jbusecke** [comment]
Sounds great! I think this is a problem worth spending some time on, because many people do this sort of thing and if we can come up with a customizable but general solution, then everybody wins.

**Comment by aaronspring** [comment]
Apparently I did this only once. I remembered doing this without the intermediate step. bias self-view.

unfortunatly, I cannot find the notebook anymore. I know I used geopandas to load the shapefile.

```python
# I just remapping this to my grid
ocean_regions = xr.open_dataset(
    "/work/mh0727/m300524/masks/ocean-provinces_0.1-0.1_GR15.nc"
).rename({"Band1": "region"})["region"]

# make only four regions
agg_regions = {
    "Southern Ocean": [50, 51, 52],
    "Tropical Pacific": [39, 40, 41],
    "North Pacific": [30, 31, 32, 34],
    "North Atlantic": [2, 3, 4],
}

def get_agg_regions(ds_sm, agg_regions=agg_regions, how="mean"):
    """Aggregate regions based on `agg_regions` for analysis from Longhurst predefined regions."""
    ds_agg_list = []
    for key, value in agg_regions.items():
        ds_agg_list.append(getattr(ds_sm.sel(region=value), how)("region"))
    ds_sm_agg = xr.concat(ds_agg_list, "region")
    ds_sm_agg["region"] = list(agg_regions.keys())
    return ds_sm_agg

**Comment by aaronspring** [comment]
finally found it: https://gist.github.com/aaronspring/dc153a64b8f1c858529b87512c2dfb24

**Comment by mathause** [comment]
Can this be closed with #91?
I'm also fine with leaving it open, but I am not sure if you are discussing how to aggregate regions or alternative datasets for ocean basins/ regions.

**Comment by jbusecke** [comment]
A bit of both, I suppose. My current needs are covered by #91, and I will implement the aggregation in [cmip6_preprocessing](https://github.com/jbusecke/cmip6_preprocessing) for now. So from my point of view this can be closed. 

Should I open a separate issue for the aggregation? 

@aaronspring if you have the bandwidth I think it would be great if we had the longhurst regions in here as well. Perhaps you can add them similarly to the natural earth data?

**Comment by aaronspring** [comment]
I can try after Thursday

**Comment by jbusecke** [comment]
Just for posteriority. I have put together a PR with the merging functionality [here](https://github.com/jbusecke/cmip6_preprocessing/pull/18). Works super nicely with CMIP6 models, see [examples here](https://github.com/jbusecke/cmip6_preprocessing/blob/regionmask_merge/doc/regionmask.ipynb).

**Comment by mathause** [comment]
Looks nice - thanks for using regionmask! The plotting issue may be resolved by setting `infer_intervals=False` in `ds.plot()`

Yes opening a new issue for the aggregation would be good. I can also do it if I find time.

I am happy to include more regions. However, this may not be as easy as I'd like it to be. There is a nice API to download naturalearth data and this won't exist for most other data sources. So one possibility is to write a generic shapefile downloader and hope the data stays in the same place. Or to include the data directly in regionmask. However, here we'd have to be careful concerning the "terms of use" and I'd also like to avoid bloating the main repo so an external data-repo may be in order (#88).

**Comment by jbusecke** [comment]
That was the reason I stayed with natural earth in the first place. If other data providers have a neat API, I think it might be worth it. 

Storing the data in a seperate repository could be useful, but yes, the terms of use might get complicated. 

I am closing this now. Lets continue discussion in #92 and potential future issues?
